engine/runtime-compiler/src/main/java/org/enso/compiler/phase/ImportResolverAlgorithm.java

```diff
@@ -0,0 +1,201 @@
+package org.enso.compiler.phase;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import org.enso.compiler.core.CompilerError;
+import org.enso.editions.LibraryName;
+import scala.Tuple2;
+
+public abstract class ImportResolverAlgorithm<
```

Comment:
Similar [coding pattern](http://wiki.apidesign.org/wiki/Singletonizer) to @radeusgd's https://github.com/enso-org/enso/pull/8652/files#diff-f506c17c22831ffcb09018cf21589de4094e513a434364b24a631ede4ce86f8eR28:
- keep the original code flow
- replace access to IR classes with generic type parameters
- trade method call on a removed class by `protected abstract` method

Then [create a subclass](https://github.com/enso-org/enso/pull/9539/files#diff-9034f9f42d0b7d9f8e8066f1ce9dfe5bf0545c3574d988825688cc3cad187aaeR23) that provides back all real types and implements the abstract methods by calling the real methods on the real IR classes.

Finally [create another subclass](https://github.com/enso-org/enso/pull/9539/files#diff-16a37c4f502f77bb7f85ed51a7a01b2ece12baf00aa72bfba5d39a4d048a3101R18) that works with completely different (runtime) classes, but _uses the same algorithm_ behind the hood.